app/forms/registration_form.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+class RegistrationForm < ApplicationForm
+  attr_accessor :email, :password, :password_confirmation, :client_id
+
+  def initialize(sign_up_params)
+    super
+    @client_app = Doorkeeper::Application.find_by(uid: sign_up_params[:client_id])
+    allowed_params = sign_up_params.except(:client_id)
+    @user = User.new(allowed_params)
+  end
+
+  def save
+    unless validate_client_id && @user.valid?
+      errors.add(:user, @user.errors.full_messages.last) unless @user.errors.empty?
+      return
+    end
+
+    @user.save
+  end
+
+  private
+
```

Comment:
We can define the `client_app` as an attr_reader like this, it's same purpose as instance variable `@` however it's clearer when we read the class. 

```
attr_reader client_app
```

In the initializer, you can init it like current approach

```
@client_app = Doorkeeper::Application.find_by(uid: sign_up_params[:client_id])
```

However in the method below, `validate_client_id`, we can ignore the `@`.

```suggestion
def validate_client_id
    return true if client_app

    errors.add(:client_id, I18n.t('doorkeeper.errors.messages.invalid_client'))
  end
```

Or even better

```ruby
  def validate_client_id
    errors.add(:client_id, I18n.t('doorkeeper.errors.messages.invalid_client')) unless client_app
  end
```